src/main/java/dev/morling/onebrc/CalculateAverage_royvanrijn.java

```diff
@@ -15,65 +15,182 @@
+import java.io.FileInputStream;
+import java.io.InputStreamReader;
+import java.io.RandomAccessFile;
+import java.nio.charset.StandardCharsets;
+import java.util.HashMap;
 import java.util.Map;
+import java.util.concurrent.ExecutorCompletionService;
+import java.util.concurrent.Executors;
 import java.util.stream.Collectors;
 
+/**
+ * Best runtimes per version:
+ *
+ * Initial submission: 62347 ms
+ * Chunked reader:     16347 ms
+ * Optimized parser:   12499 ms
+ * Branchless methods: 10514 ms
+ *
+ * Best performance on MacBook M2 Pro: 21.0.1-graal
+ * `sdk use java 21.0.1-graal`
+ */
 public class CalculateAverage_royvanrijn {
-
     private static final String FILE = "./measurements.txt";
+    private static final int CHUNKS = 16;
 
-    private record Measurement(double min, double max, double sum, long count) {
+    // mutable state now instead of records, ugh, less instantiation.
+    static final class Measurement {
+        int min, max, count;
+        long sum;
 
-        Measurement(double initialMeasurement) {
-            this(initialMeasurement, initialMeasurement, initialMeasurement, 1);
+        public Measurement() {
+            this.min = 10000;
+            this.max = -10000;
         }
 
-        public static Measurement combineWith(Measurement m1, Measurement m2) {
-            return new Measurement(
-                    m1.min < m2.min ? m1.min : m2.min,
-                    m1.max > m2.max ? m1.max : m2.max,
-                    m1.sum + m2.sum,
-                    m1.count + m2.count
-            );
+        public Measurement(int min, int max, long sum, int count) {
+            this.min = min;
+            this.max = max;
+            this.sum = sum;
+            this.count = count;
+        }
+
+        public Measurement updateWith(int measurement) {
+            min = min(min, measurement);
+            max = max(max, measurement);
+            sum += measurement;
+            count++;
+            return this;
+        }
+
+        public Measurement updateWith(Measurement measurement) {
+            min = min(min, measurement.min);
+            max = max(max, measurement.max);
+            sum += measurement.sum;
+            count += measurement.count;
+            return this;
         }
 
         public String toString() {
-            return round(min) + "/" + round(sum / count) + "/" + round(max);
+            return round(min) + "/" + round((1.0 * sum) / count) + "/" + round(max);
         }
 
         private double round(double value) {
-            return Math.round(value * 10.0) / 10.0;
+            return Math.round(value) / 10.0;
+        }
+    }
+
+    public static final void main(String[] args) throws Exception {
+
+        // Split the file into numberOfChunks, use random access to read and process in parallel.
+
+        final long fileSize = new File(FILE).length();
+        final long chunkSize = fileSize / CHUNKS;
+
+        final var es = Executors.newVirtualThreadPerTaskExecutor();
```

Comment:
I think VT are not well suited for file IO. Have you tried a regular executor? In my case it is faster.